corgi/tasks/brew.py

```diff
@@ -698,3 +699,82 @@ def slow_refresh_brew_build_tags(build_id: int) -> None:
     logger.info(f"Finished refreshing Brew build tags for {build_id}")
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def slow_delete_brew_build(build_id: int, build_state: int) -> int:
+    """Delete a Brew build (and its relations) in Corgi when it's deleted in Brew"""
+    # Shipped builds (which have a Brew tag for any product) are never deleted
+    # Only unshipped builds not part of any product are deleted
+    # For example, builds that failed QE will not become part of a product
+    # We don't need this data for any of our use cases
+    # Keeping it causes other issues when reloading old builds, scanning for licenses, etc.
+
+    logger.info(f"Deleting Brew build {build_id} with state {build_state}")
+    if build_state != koji.BUILD_STATES["DELETED"]:
+        raise ValueError(
+            f"Invalid state for build {build_id}: "
+            f"expected {koji.BUILD_STATES['DELETED']}, received {build_state}"
+        )
+
+    deleted_count = 0
+    with transaction.atomic():
+        # Build might not exist, or might exist but have no / multiple root components
+        # TODO: Check this! Why do builds have multiple root components?
+        root_components_list = (
+            SoftwareBuild.objects.filter(build_id=build_id, build_type=SoftwareBuild.Type.BREW)
+            .exclude(components__isnull=True)
+            .values_list("components", flat=True)
+            or ()
+        )
+        for root_component_pk in root_components_list:
+            # If the build has root components, delete the child components that are
+            # provided by / upstreams of only this build's root, and not any other builds
+            # Deleting the Component will automatically delete the ComponentNodes
+            provided_components = Component.objects.annotate(sources_count=Count("sources")).filter(
+                sources=root_component_pk, sources_count=1
+            )
+            deleted_count += _delete_queryset(provided_components, "provided component")
+
+            upstream_components = (
+                # Doing .filter(downstreams=root_component_pk)
+                # before .annotate(downstreams_count=Count("downstreams"))
+                # makes Django return the wrong results
+                # https://docs.djangoproject.com/en/3.2/topics/db/aggregation/
+                # #order-of-annotate-and-filter-clauses
+                Component.objects.annotate(downstreams_count=Count("downstreams")).filter(
+                    downstreams=root_component_pk, downstreams_count=1
+                )
+            )
+            deleted_count += _delete_queryset(upstream_components, "upstream component")
+
+        # Relations without a linked (NULL) build are "unprocessed"
+        # We process these relations each day by loading their build ID
+        # Deleting the relation avoids reloading the build we are deleting
+        relations = ProductComponentRelation.objects.filter(
```

Comment:
Nitpick: Once we update the ERRATA relations to only be created for shipped errata (CORGI-723) we should never have a deleted build in the relations table. Perhaps it's a good idea to fail the task in this case so we can investigate why that is happening. 

I'm happy for this to be merged as is, and we can update the task to fail in this case once CORGI-723 is implemented.